chainer/optimizer.py

```diff
@@ -27,153 +30,314 @@ def exponential_decay_noise(xp, shape, dtype, hook, opt):
-        method or CPU/GPU dedicated versions (:meth:`init_state_cpu` and
-        :meth:`init_state_gpu`).
+        del self._hooks[name]
+
+    def update(self, param):
+        """Invokes hook functions and updates the parameter.
 
         Args:
-            param (~chainer.Variable): Parameter variable.
-            state (dict): State dictionary.
+            param (~chainer.Variable): Variable to be updated.
+
+        """
+        if not self.enabled:
+            return
+
+        self.t += 1
+        self._prepare(param)
+        for hook in six.itervalues(self._hooks):
+            hook(self, param)
+        self.update_core(param)
+
+    def update_core(self, param):
+        """Updates the parameter.
+
+        Implementation of UpdateRule should override this method or both of
+        :meth:`_update_core_cpu` and :meth:`_update_core_gpu`.
 
-        .. seealso:: :meth:`init_state_cpu`, :meth:`init_state_gpu`
+        Args:
+            param (~chainer.Variable): Variable to be updated.
 
         """
         with cuda.get_device(param.data) as dev:
             if int(dev) == -1:
-                self.init_state_cpu(param, state)
+                self.update_core_cpu(param)
             else:
-                self.init_state_gpu(param, state)
+                self.update_core_gpu(param)
 
-    def init_state_cpu(self, param, state):
-        """Initializes the optimizer state on CPU.
+    def update_core_cpu(self, param):
+        """Updates the parameter on CPU.
 
-        This method is called from :meth:`init_state` by default.
+        See :meth:`update_core` for details.
 
         Args:
-            param (~chainer.Variable): Parameter variable. Its data array is
-                of type :class:`numpy.ndarray`.
-            state (dict): State dictionary.
+            param (~chainer.Variable): Variable to be updated.
+
+        """
+        raise NotImplementedError
+
+    def update_core_gpu(self, param):
+        """Updates the parameter on GPU.
+
+        See :meth:`update_core` for details.
 
-        .. seealso:: :meth:`init_state`
+        Args:
+            param (~chainer.Variable): Variable to be updated.
+
+        """
+        raise NotImplementedError
+
+    def init_state(self, param):
+        """Initializes the state.
+
+        Any implementations that use the state should override this mehtod.
+        This method is called at the first update.
+
+        Args:
+            param (~chainer.Variable): Parameter variable. It can be used to
+                extract the shape and the data type of the parameter.
 
         """
         pass
 
-    def init_state_gpu(self, param, state):
-        """Initializes the optimizer state on GPU.
+    def serialize(self, serializer):
+        """Serializes the update rule state.
 
-        This method is called from :meth:`init_state` by default.
+        Be careful that this method only saves/loads the state of the update
+        rule.
```

Comment:
The parameters of the target link is not saved/loaded by this method. I'll add a note.